packages/query/src/block-processor.ts

```diff
@@ -0,0 +1,390 @@
+        'updated commitments': commitmentRecords.size,
+        'new notes': compactScan?.newNotes.length,
+        'new swaps': compactScan?.newSwaps.length,
+      };
+
+      if (Object.values(fullScanReasons).some(Boolean)) {
+        console.log('full scan', compactBlock.height, fullScanReasons);
+
+        // this is a network query
+        const fullBlock = await this.querier.app.txsByHeight(compactBlock.height);
+
+        const relevantTx = await this.identifyRelevantTransactions(
+          Array.from(spentNullifiers.keys()),
+          Array.from(commitmentRecords.keys()),
+          fullBlock.transactions,
+        );
+
+        console.warn('relevantTx', relevantTx.length);
+
+        // during wasm tx info generation later, wasm independently queries
+        // database for asset metadata, so we have to pre-populate. LpNft
+        // position states aren't known by the chain so aren't poulated by
+        // identifyNewAssets
+        // - detect LpNft position opens, and generate all possible position state metadata
+        // - update idb
+        await this.identifyLpNftPositions(fullBlock.transactions);
+
+        // at this point txinfo can be generated and saved. this will resolve
+        // pending broadcasts, and populate the transaction list.
+        // - calls wasm for each relevant tx
+        // - saves to idb
+        await this.saveTransactionInfo(compactBlock.height, relevantTx);
+      }
+    }
+  }
+
+  private async identifyNewAssets(newNotes: SpendableNoteRecord[]) {
+    for (const n of newNotes) {
+      const assetId = n.note?.value?.assetId;
+      if (!assetId) continue;
+      if (await this.indexedDb.getAssetsMetadata(assetId)) continue;
+
+      let metadata;
+      metadata = await this.querier.shieldedPool.denomMetadata(assetId);
+
+      if (!metadata) {
+        const UNNAMED_ASSET_PREFIX = 'passet';
+        const denom = bech32.encode(UNNAMED_ASSET_PREFIX, bech32.toWords(assetId.inner));
+        metadata = new DenomMetadata({
+          base: denom,
+          denomUnits: [{ aliases: [], denom, exponent: 0 }],
+          display: denom,
+          penumbraAssetId: assetId,
+        });
+      }
+
+      await this.indexedDb.saveAssetsMetadata(metadata);
+    }
+  }
+
+  private async resolveCommitments(commitments?: StoreCommitment[]) {
+    const records = new Map<StateCommitment, SwapRecord | SpendableNoteRecord>();
+
+    // - match commitments to idb notes/swaps
+    for (const storeCommitment of commitments ?? []) {
+      const stateCommitment = StateCommitment.fromJson(storeCommitment.commitment);
+      const record =
+        (await this.indexedDb.getSpendableNoteByCommitment(stateCommitment)) ??
+        (await this.indexedDb.getSwapByCommitment(stateCommitment));
+
+      if (!record) continue;
+
+      records.set(stateCommitment, record);
+    }
+
+    if (commitments?.length) console.log('resolveCommitments', commitments.length, records.size);
+
+    return records;
+  }
```

Comment:
This is a very suboptimal way to get commitments from the current block
The point is that `flush_updates() `returns `new_notes` and `new_swaps` where commitments are defined. 
Instead of using `sctUpdates.store_commitments` , get notes and swap records from indexed-db we can just use commitments from `new_swaps` and `new_notes`

Reply:
as far as i can tell, sctUpdates.store_commitments is equivalent to commitments in new_notes and new_swaps.

i am now using commitments from new_swaps and new_notes, but do you have an example or can you explain where store_commitments is suboptimal?

Reply:
We can avoid additional indexed-db calls 
```
const record =
        (await this.indexedDb.getSpendableNoteByCommitment(stateCommitment))) ??
        (await this.indexedDb.getSwapByCommitment(stateCommitment));

      if (!record) continue;
```

Reply:
ok 👍 but that was written not believing that the lists are equivalent - will store_commitments ever contain something we don't have? primarily i'm getting this list to identify relevant tx. if we can just use store_commitments directly to scan tx this is better than both approaches

Reply:
I'll take some time to test the latest changes of block-processor and get back to  discussion/review